pkg/ls/ingestdoc.go

```diff
@@ -101,18 +102,49 @@ func (i ingestCursor) getOutput() *lpg.Node {
 	return i.output[len(i.output)-1]
 }
 
-func Ingest(builder GraphBuilder, root ParsedDocNode) (*lpg.Node, error) {
+type Ingester struct {
+	Schema *Layer
+	cached map[*lpg.Node]struct{}
+	mu     sync.RWMutex
+}
+
+func (ing *Ingester) Ingest(builder GraphBuilder, root ParsedDocNode) (*lpg.Node, error) {
 	cursor := ingestCursor{
 		input: []ParsedDocNode{root},
 	}
 	_, n, err := ingestWithCursor(builder, cursor)
 	if err != nil {
 		return n, err
 	}
-	err = builder.PostIngest(root.GetSchemaNode(), n)
+	nodeIDMap := GetSchemaNodeIDMap(n)
+	ing.cached = make(map[*lpg.Node]struct{})
+	ing.GetPostIngestSchemaNodes(root.GetSchemaNode())
+	for schemaNode := range ing.cached {
+		builder.PostIngestSchemaNode(root.GetSchemaNode(), schemaNode, n, nodeIDMap)
+	}
 	return n, err
 }
 
+func (ing *Ingester) GetPostIngestSchemaNodes(schemaRootNode *lpg.Node) {
+	ing.mu.Lock()
+	ForEachAttributeNode(schemaRootNode, func(schemaNode *lpg.Node, _ []*lpg.Node) bool {
```

Comment:
Return the cached list if it is initialized. Intitialize the cache if it is not initialized.

Use RLock if the cache is already initialized.,